# Homework 5
## Ziyan Hu

## Problem 1

In [2]:
import pandas as pd
import numpy as np

In [70]:
# read the csv
listings_df = pd.read_csv('listings.csv')
listings_df.head()

id                         listing_url       scrape_id last_scraped  \
0   3781   https://www.airbnb.com/rooms/3781  20230916060056   2023-09-16   
1   5506   https://www.airbnb.com/rooms/5506  20230916060056   2023-09-16   
2   6695   https://www.airbnb.com/rooms/6695  20230916060056   2023-09-16   
3   8789   https://www.airbnb.com/rooms/8789  20230916060056   2023-09-16   
4  10813  https://www.airbnb.com/rooms/10813  20230916060056   2023-09-16   

        source                                               name  \
0  city scrape  Rental unit in Boston · ★4.96 · 1 bedroom · 1 ...   
1  city scrape  Guest suite in Boston · ★4.79 · 1 bedroom · 1 ...   
2  city scrape  Condo in Boston · ★4.80 · Studio · 2 beds · 1 ...   
3  city scrape  Rental unit in Boston · ★4.67 · 1 bedroom · 1 ...   
4  city scrape  Rental unit in Boston · ★5.0 · Studio · 1 bed ...   

                                         description  \
0  Fully separate apartment in a two apartment bu...   
1  Private guest room with private bath, You do n...   
2  Comfortable,  Fully Equipped private apartment...   
3  This unit is for sale. There will need to be o...   
4  Stunning Back Bay furnished studio apartment. ...   

                               neighborhood_overview  \
0  Mostly quiet ( no loud music, no crowed sidewa...   
1  Peaceful, Architecturally interesting, histori...   
2  Peaceful, Architecturally interesting, histori...   
3  Beacon Hill is a historic neighborhood filled ...   
4  Wander around this quintessential neighborhood...   

                                         picture_url  host_id  ...  \
0  https://a0.muscache.com/pictures/24670/b2de044...     4804  ...   
1  https://a0.muscache.com/pictures/miso/Hosting-...     8229  ...   
2  https://a0.muscache.com/pictures/38ac4797-e7a4...     8229  ...   
3  https://a0.muscache.com/pictures/miso/Hosting-...    26988  ...   
4  https://a0.muscache.com/pictures/20b5b9c9-e1f4...    38997  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        4.96                   4.88                4.92   
1                        4.88                   4.54                4.74   
2                        4.96                   4.51                4.72   
3                        4.96                   4.96                4.56   
4                        5.00                   5.00                4.75   

                      license instant_bookable calculated_host_listings_count  \
0                         NaN                f                              1   
1  Approved by the government                f                             10   
2                   STR446650                f                             10   
3                         NaN                f                              8   
4                         NaN                f                             11   

  calculated_host_listings_count_entire_homes  \
0                                           1   
1                                          10   
2                                          10   
3                                           8   
4                                          11   

  calculated_host_listings_count_private_rooms  \
0                                            0   
1                                            0   
2                                            0   
3                                            0   
4                                            0   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.24  
1                                           0              0.67  
2                                           0              0.73  
3                                           0              0.24  
4                                           0              0.06  

[5 rows x 75 columns]

In [71]:
#1a
entries = len(listings_df)
identifiers = listings_df['id'].nunique()
entries,identifiers

(4033, 4033)

There are 4033 entries and 4033 identifiers for the Boston AirBnB

In [72]:
#1b
reviews_df = pd.read_csv('reviews.csv')
reviews_df.head()
#the number of reviews are ...
num_reviews = len(reviews_df)
#unique identifiers of real reviews...
unique_id = reviews_df['listing_id'].nunique()
num_reviews,unique_id

(184680, 3058)

There are 184680 entries in reviews.csv and 3058 unique identifiers.

In [73]:
#1b cont.
# ... listings are reviewed in this table (not all 4033)
listings_have_reviews = listings_df['id'].isin(reviews_df['listing_id'])
listings_have_reviews

0        True
1        True
2        True
3        True
4        True
        ...  
4028    False
4029     True
4030    False
4031    False
4032    False
Name: id, Length: 4033, dtype: bool

From the table above we can see that not all the listings have a corresponding review.

In [74]:
#1c
# merge 2 data frames
merged_df = pd.merge(listings_df, reviews_df,
                     how='left',
                     left_on='id',
                     right_on='listing_id')

# Calculate the number of rows in the merged DataFrame
num_rows_merged = len(merged_df)

# Calculate the number of unique AirBnB IDs in the merged DataFrame
unique_ids_merged = merged_df['id'].nunique()

# ... rows of listings, ... of those () with NaN review
nan_filled_rows = merged_df['listing_id'].isna().sum()

num_rows_merged,unique_ids_merged,nan_filled_rows

(185655, 4033, 975)

We have 185655 rows and 4033 unique AirBnB ids and 975 rows were filled in with NaN

In [75]:
#1d
# Merge the listings and reviews DataFrames using an inner merge
inner_merged_df = pd.merge(listings_df, reviews_df, how='inner', left_on='id', right_on='listing_id')

# Calculate the number of rows in the inner merged DataFrame
num_rows = len(inner_merged_df)

# Calculate the number of unique AirBnB IDs in the inner merged DataFrame
unique_ids = inner_merged_df['id'].nunique()

num_rows, unique_ids


(184680, 3058)

There are 184680 rows and 3058 unique AirBnB ids

In [4]:
#1e
# read the full listing csv
full_listings_df = pd.read_csv('listings.csv.gz',
                          compression='gzip')
full_listings_df.head()

id                         listing_url       scrape_id last_scraped  \
0   3781   https://www.airbnb.com/rooms/3781  20230916060056   2023-09-16   
1   5506   https://www.airbnb.com/rooms/5506  20230916060056   2023-09-16   
2   6695   https://www.airbnb.com/rooms/6695  20230916060056   2023-09-16   
3   8789   https://www.airbnb.com/rooms/8789  20230916060056   2023-09-16   
4  10813  https://www.airbnb.com/rooms/10813  20230916060056   2023-09-16   

        source                                               name  \
0  city scrape  Rental unit in Boston · ★4.96 · 1 bedroom · 1 ...   
1  city scrape  Guest suite in Boston · ★4.79 · 1 bedroom · 1 ...   
2  city scrape  Condo in Boston · ★4.80 · Studio · 2 beds · 1 ...   
3  city scrape  Rental unit in Boston · ★4.67 · 1 bedroom · 1 ...   
4  city scrape  Rental unit in Boston · ★5.0 · Studio · 1 bed ...   

                                         description  \
0  Fully separate apartment in a two apartment bu...   
1  Private guest room with private bath, You do n...   
2  Comfortable,  Fully Equipped private apartment...   
3  This unit is for sale. There will need to be o...   
4  Stunning Back Bay furnished studio apartment. ...   

                               neighborhood_overview  \
0  Mostly quiet ( no loud music, no crowed sidewa...   
1  Peaceful, Architecturally interesting, histori...   
2  Peaceful, Architecturally interesting, histori...   
3  Beacon Hill is a historic neighborhood filled ...   
4  Wander around this quintessential neighborhood...   

                                         picture_url  host_id  ...  \
0  https://a0.muscache.com/pictures/24670/b2de044...     4804  ...   
1  https://a0.muscache.com/pictures/miso/Hosting-...     8229  ...   
2  https://a0.muscache.com/pictures/38ac4797-e7a4...     8229  ...   
3  https://a0.muscache.com/pictures/miso/Hosting-...    26988  ...   
4  https://a0.muscache.com/pictures/20b5b9c9-e1f4...    38997  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        4.96                   4.88                4.92   
1                        4.88                   4.54                4.74   
2                        4.96                   4.51                4.72   
3                        4.96                   4.96                4.56   
4                        5.00                   5.00                4.75   

                      license instant_bookable calculated_host_listings_count  \
0                         NaN                f                              1   
1  Approved by the government                f                             10   
2                   STR446650                f                             10   
3                         NaN                f                              8   
4                         NaN                f                             11   

  calculated_host_listings_count_entire_homes  \
0                                           1   
1                                          10   
2                                          10   
3                                           8   
4                                          11   

  calculated_host_listings_count_private_rooms  \
0                                            0   
1                                            0   
2                                            0   
3                                            0   
4                                            0   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.24  
1                                           0              0.67  
2                                           0              0.73  
3                                           0              0.24  
4                                           0              0.06  

[5 rows x 75 columns]

In [77]:
#1e cont. (How many listings by superhost)
full_listings_df[full_listings_df['host_is_superhost'] == 't']['id'].nunique()

1014

In [78]:
#1e cont. (How many unique superhosts)
full_listings_df[full_listings_df['host_is_superhost'] == 't']['host_id'].nunique()

490

In [79]:
#1e cont. (How many superhost with no reviews on any listings)
# Merge the superhosts from full_listings_df with the reviews dataframe
superhost_and_reviews = pd.merge(full_listings_df[full_listings_df['host_is_superhost'] == 't'],
                                 reviews_df,
                                 how='left',
                                 left_on='id',
                                 right_on='listing_id')

# Filter superhosts who don't have any reviews (where review 'listing_id' is NaN)
superhosts_without_reviews = superhost_and_reviews[
    (superhost_and_reviews['host_is_superhost'] == 't') &
    (superhost_and_reviews['listing_id'].isna())
    ]

# Find the number of unique superhosts without reviews
unique_superhosts_without_reviews = superhosts_without_reviews['host_id'].nunique()

unique_superhosts_without_reviews


22

There are 22 of unique superhosts without reviews

## Problem 2

In [80]:
citibike = pd.read_csv('citibike_tripdata_2023-09.csv')
citibike.head()

ride_id  rideable_type     started_at       ended_at  \
0  36816E453063E92A   classic_bike  9/12/23 15:56  9/12/23 16:03   
1  5737471912BBD88F   classic_bike  9/12/23 18:16  9/12/23 18:57   
2  B014B020DC96CA66  electric_bike  9/11/23 14:28  9/11/23 14:35   
3  46F26D7BE22C25CA   classic_bike   9/17/23 9:03   9/17/23 9:06   
4  74872D06EA677BFA   classic_bike   9/16/23 2:42   9/16/23 3:16   

     start_station_name start_station_id       end_station_name  \
0      5 Ave & E 103 St          7511.19        E 88 St & 1 Ave   
1    E 14 St & Avenue B          5736.09  West St & Chambers St   
2       W 24 St & 7 Ave          6257.03  Lafayette St & E 8 St   
3  Center Blvd & 51 Ave          6179.07   Vernon Blvd & 50 Ave   
4      W 52 St & 11 Ave          6926.01    Broadway & Kent Ave   

  end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  
0        7235.13  40.792197 -73.952660  40.778301 -73.948813        member  
1        5329.03  40.729387 -73.977724  40.717548 -74.013221        member  
2        5788.13  40.744876 -73.995299  40.730207 -73.991026        member  
3        6170.02  40.743366 -73.959607  40.742327 -73.954117        member  
4        5134.09  40.767419 -73.994259  40.710888 -73.968329        member

### (a)

In [81]:
# Calculate the number of rows in the citibike dataframe
num_rows = len(citibike)
num_rows

100000

In [82]:
# Calculate the number of unique stations
unique_stations = pd.concat([citibike['start_station_name'], citibike['end_station_name']]).nunique()
unique_stations

1961

There are 100000 citibike rides in this dataset, and 1962 unique stations were visited

### (b)

In [83]:
def count_bike_type(df, member_or_casual):
    return df[df['member_casual'] == member_or_casual].shape[0]

# Group by rideable_type and apply the function
member_count = citibike.groupby('rideable_type').apply(count_bike_type, member_or_casual='member')
casual_count = citibike.groupby('rideable_type').apply(count_bike_type, member_or_casual='casual')


print('member count \n', member_count)
print()
print('casual count \n', casual_count)

member count 
 rideable_type
classic_bike     73945
electric_bike     5449
dtype: int64

casual count 
 rideable_type
classic_bike     18989
electric_bike     1617
dtype: int64


Members took 73,945 classic bike rides and 5,449 electric bike rides. On the other hand, casual users took 18,989 classic bike rides and 1,617 electric bike rides in September 2023.

### (c)

In [84]:
# Convert started_at and ended_at columns to datetime
citibike['started_at'] = pd.to_datetime(citibike['started_at'])
citibike['ended_at'] = pd.to_datetime(citibike['ended_at'])

# Calculate ride duration in minutes
citibike['ride_duration'] = (citibike['ended_at'] - citibike['started_at']).dt.total_seconds() / 60

# Calculate the overall mean ride duration
overall_mean_duration = citibike['ride_duration'].mean()
overall_mean_duration

16.0426

In [85]:
# Calculate mean ride duration for electric and classic rides
mean_ride_time_classic = citibike[citibike['rideable_type'] == 'classic_bike']['ride_duration'].mean()
mean_ride_time_electric = citibike[citibike['rideable_type'] == 'electric_bike']['ride_duration'].mean()
(mean_ride_time_classic,mean_ride_time_electric)

(16.119547205543718, 15.030568921596377)

Therefore, mean duration is 16.0426 minutes overall. For electric bike, mean duration is 15.030569 minutes, while for classic bike mean duration is 16.119547 minutes. This difference is probably because electric bikes are faster than classic, so rides of a similar distance would take less time.

### (d)

In [86]:
# Filter rides that started at "Amsterdam Ave & W 79 St" and count unique end stations
start_amsterdam = citibike[citibike['start_station_name'] == 'Amsterdam Ave & W 79 St']
print(start_amsterdam['end_station_name'].nunique())

105


There are 105 different stations for the rides were started at Amsterdam Ave & 79 St

### (e)

In [87]:
# Count the number of rides that started and ended at the same station
len(citibike[citibike['start_station_name'] == citibike['end_station_name']])

5125

There are 5125 rides start and end at the same station

### (f)

In [88]:
# Define the function to calculate the Euclidean distance
def distance(start_lat, start_lng, end_lat, end_lng):
    lat_diff = start_lat - end_lat
    lng_diff = start_lng - end_lng
    return (lat_diff**2 + lng_diff**2)**0.5

# Apply the distance function to the dataframe
citibike['ride_distance'] = citibike.apply(lambda row: distance(row['start_lat'], row['start_lng'], row['end_lat'], row['end_lng']), axis=1)

In [89]:
print(citibike.groupby('member_casual')['ride_distance'].mean())

member_casual
casual    0.021927
member    0.019058
Name: ride_distance, dtype: float64


Casual users ride longer distances on average.